### ISSUE TITLE
context bounds not allowed in algebras

### ISSUE THREAD

**Issue by fommil** [47deg/freestyle#365]
I had this algebra that worked in freestyle 0.2

```scala
  final case class Response[T](header: HttpResponseHeader, body: T)

  @free trait JsonHttpClient {
    def get[B: Decoder](uri: Uri,
                        headers: List[HttpHeader] = Nil): FS[Response[B]]

    // using application/x-www-form-urlencoded
    def postUrlencoded[A: UrlEncoded, B: Decoder](
      uri: Uri,
      payload: A,
      headers: List[HttpHeader] = Nil
    ): FS[Response[B]]
  }
```

the idea being that it can be a simple JSON HTTP client that takes in a circe `Decoder` for responses and responds as domain objects... so the caller can say what they expect and get it.

However, in freestyle 0.3 this fails with a macro error regarding `Decoder` implicit divergence (which I think is a red herring). Converting the context bound notation to use an explicit `implicit` parameter block, I get this warning

```
case classes should have a non-implicit parameter list; adapting to 'case class PostUrlencodedOP()(...)'
```

followed by this error

```
[error] <macro>:27: no arguments allowed for nullary method apply
```

which looks like implicits are just not being handled.

It's very annoying that `scala.meta` doesn't output the location of where the macro is being applied in terms of the file that generated it, because it can be super hard to localise where problems are coming from. @xeno-by can anything be done about that?

By explicit implicits I mean

```scala
  @free trait JsonHttpClient {
    def get[B](
      uri: Uri,
      headers: List[HttpHeader] = Nil
    )(
      implicit decoder: Decoder[B]
    ): FS[Response[B]]

    // using application/x-www-form-urlencoded
    def postUrlencoded[A, B](
      uri: Uri,
      payload: A,
      headers: List[HttpHeader] = Nil
    )(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]]
  }
```

**Comment by suhasgaddam** [comment]
https://github.com/frees-io/freestyle/pull/360 to debug the expanded macro in `0.3.1-SNAPSHOT`

~~I have a feeling this is related to flatten on the parameter list.~~

**Comment by fommil** [comment]
I've pushed the code to https://github.com/fommil/drone-dynamic-agents/tree/freestyle-bug just run `sbt compile`

with context bounds

```
{
  trait JsonHttpClient[FF$24[_]] extends _root_.freestyle.internal.EffectLike[FF$24] {
    def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): FS[Response[B]]
    def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): FS[Response[B]]
  }
  object JsonHttpClient {
    sealed trait Op[_] extends scala.Product with java.io.Serializable { val FSAlgebraIndex23: _root_.scala.Int }
    case class GetOP[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 0 }
    case class PostUrlencodedOP[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 1 }
    type OpTypes = _root_.iota.KCons[Op, _root_.iota.KNil]
    trait Handler[MM$30[_]] extends _root_.freestyle.FSHandler[Op, MM$30] {
      protected[this] def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): MM$30[Response[B]]
      protected[this] def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): MM$30[Response[B]]
      override def apply[AA$31](fa$32: Op[AA$31]): MM$30[AA$31] = ((fa$32.FSAlgebraIndex23: @_root_.scala.annotation.switch) match {
        case 0 =>
          val fresh33: GetOP[Any] = fa$32.asInstanceOf[GetOP[Any]]
          get(fresh33.uri, fresh33.headers)
        case 1 =>
          val fresh34: PostUrlencodedOP[Any, Any] = fa$32.asInstanceOf[PostUrlencodedOP[Any, Any]]
          postUrlencoded(fresh34.uri, fresh34.payload, fresh34.headers)
        case i =>
          throw new _root_.java.lang.Exception(s"freestyle internal error: index " + i + " out of bounds for " + this)
      }).asInstanceOf[MM$30[AA$31]]
    }
    class To[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]) extends JsonHttpClient[LL$26] {
      private[this] val toInj28 = _root_.freestyle.FreeS.inject[Op, LL$26](ii$27)
      override def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): FS[Response[B]] = toInj28(GetOP[B](uri, headers))
      override def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): FS[Response[B]] = toInj28(PostUrlencodedOP[A, B](uri, payload, headers))
    }
    implicit def to[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]): To[LL$26] = new To[LL$26]
    def apply[LL$26[_]](implicit ev$29: JsonHttpClient[LL$26]): JsonHttpClient[LL$26] = ev$29
  }
}
```

with explicit implicit args

```
{
  trait JsonHttpClient[FF$24[_]] extends _root_.freestyle.internal.EffectLike[FF$24] {
    def get[B](uri: Uri, headers: List[HttpHeader] = Nil)(implicit decoder: Decoder[B]): FS[Response[B]]
    def postUrlencoded[A, B](uri: Uri, payload: A, headers: List[HttpHeader] = Nil)(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]]
  }
  object JsonHttpClient {
    sealed trait Op[_] extends scala.Product with java.io.Serializable { val FSAlgebraIndex23: _root_.scala.Int }
    case class GetOP[B](implicit uri: Uri, headers: List[HttpHeader] = Nil, decoder: Decoder[B]) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 0 }
    case class PostUrlencodedOP[A, B](implicit uri: Uri, payload: A, headers: List[HttpHeader] = Nil, encoder: UrlEncoded[A], decoder: Decoder[B]) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 1 }
    type OpTypes = _root_.iota.KCons[Op, _root_.iota.KNil]
    trait Handler[MM$30[_]] extends _root_.freestyle.FSHandler[Op, MM$30] {
      protected[this] def get[B](implicit uri: Uri, headers: List[HttpHeader] = Nil, decoder: Decoder[B]): MM$30[Response[B]]
      protected[this] def postUrlencoded[A, B](implicit uri: Uri, payload: A, headers: List[HttpHeader] = Nil, encoder: UrlEncoded[A], decoder: Decoder[B]): MM$30[Response[B]]
      override def apply[AA$31](fa$32: Op[AA$31]): MM$30[AA$31] = ((fa$32.FSAlgebraIndex23: @_root_.scala.annotation.switch) match {
        case 0 =>
          val fresh33: GetOP[Any] = fa$32.asInstanceOf[GetOP[Any]]
          get(fresh33.uri, fresh33.headers, fresh33.decoder)
        case 1 =>
          val fresh34: PostUrlencodedOP[Any, Any] = fa$32.asInstanceOf[PostUrlencodedOP[Any, Any]]
          postUrlencoded(fresh34.uri, fresh34.payload, fresh34.headers, fresh34.encoder, fresh34.decoder)
        case i =>
          throw new _root_.java.lang.Exception(s"freestyle internal error: index " + i + " out of bounds for " + this)
      }).asInstanceOf[MM$30[AA$31]]
    }
    class To[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]) extends JsonHttpClient[LL$26] {
      private[this] val toInj28 = _root_.freestyle.FreeS.inject[Op, LL$26](ii$27)
      override def get[B](uri: Uri, headers: List[HttpHeader] = Nil)(implicit decoder: Decoder[B]): FS[Response[B]] = toInj28(GetOP[B](uri, headers, decoder))
      override def postUrlencoded[A, B](uri: Uri, payload: A, headers: List[HttpHeader] = Nil)(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]] = toInj28(PostUrlencodedOP[A, B](uri, payload, headers, encoder, decoder))
    }
    implicit def to[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]): To[LL$26] = new To[LL$26]
    def apply[LL$26[_]](implicit ev$29: JsonHttpClient[LL$26]): JsonHttpClient[LL$26] = ev$29
  }
}
```

**Comment by fommil** [comment]
```diff
--- context	2017-06-20 18:44:31.333488630 +0100
+++ explicit	2017-06-20 18:44:40.700285765 +0100
@@ -1,31 +1,31 @@
 {
   trait JsonHttpClient[FF$24[_]] extends _root_.freestyle.internal.EffectLike[FF$24] {
-    def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): FS[Response[B]]
-    def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): FS[Response[B]]
+    def get[B](uri: Uri, headers: List[HttpHeader] = Nil)(implicit decoder: Decoder[B]): FS[Response[B]]
+    def postUrlencoded[A, B](uri: Uri, payload: A, headers: List[HttpHeader] = Nil)(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]]
   }
   object JsonHttpClient {
     sealed trait Op[_] extends scala.Product with java.io.Serializable { val FSAlgebraIndex23: _root_.scala.Int }
-    case class GetOP[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 0 }
-    case class PostUrlencodedOP[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 1 }
+    case class GetOP[B](implicit uri: Uri, headers: List[HttpHeader] = Nil, decoder: Decoder[B]) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 0 }
+    case class PostUrlencodedOP[A, B](implicit uri: Uri, payload: A, headers: List[HttpHeader] = Nil, encoder: UrlEncoded[A], decoder: Decoder[B]) extends AnyRef with Op[Response[B]] { override val FSAlgebraIndex23: _root_.scala.Int = 1 }
     type OpTypes = _root_.iota.KCons[Op, _root_.iota.KNil]
     trait Handler[MM$30[_]] extends _root_.freestyle.FSHandler[Op, MM$30] {
-      protected[this] def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): MM$30[Response[B]]
-      protected[this] def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): MM$30[Response[B]]
+      protected[this] def get[B](implicit uri: Uri, headers: List[HttpHeader] = Nil, decoder: Decoder[B]): MM$30[Response[B]]
+      protected[this] def postUrlencoded[A, B](implicit uri: Uri, payload: A, headers: List[HttpHeader] = Nil, encoder: UrlEncoded[A], decoder: Decoder[B]): MM$30[Response[B]]
       override def apply[AA$31](fa$32: Op[AA$31]): MM$30[AA$31] = ((fa$32.FSAlgebraIndex23: @_root_.scala.annotation.switch) match {
         case 0 =>
           val fresh33: GetOP[Any] = fa$32.asInstanceOf[GetOP[Any]]
-          get(fresh33.uri, fresh33.headers)
+          get(fresh33.uri, fresh33.headers, fresh33.decoder)
         case 1 =>
           val fresh34: PostUrlencodedOP[Any, Any] = fa$32.asInstanceOf[PostUrlencodedOP[Any, Any]]
-          postUrlencoded(fresh34.uri, fresh34.payload, fresh34.headers)
+          postUrlencoded(fresh34.uri, fresh34.payload, fresh34.headers, fresh34.encoder, fresh34.decoder)
         case i =>
           throw new _root_.java.lang.Exception(s"freestyle internal error: index " + i + " out of bounds for " + this)
       }).asInstanceOf[MM$30[AA$31]]
     }
     class To[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]) extends JsonHttpClient[LL$26] {
       private[this] val toInj28 = _root_.freestyle.FreeS.inject[Op, LL$26](ii$27)
-      override def get[B: Decoder](uri: Uri, headers: List[HttpHeader] = Nil): FS[Response[B]] = toInj28(GetOP[B](uri, headers))
-      override def postUrlencoded[A: UrlEncoded, B: Decoder](uri: Uri, payload: A, headers: List[HttpHeader] = Nil): FS[Response[B]] = toInj28(PostUrlencodedOP[A, B](uri, payload, headers))
+      override def get[B](uri: Uri, headers: List[HttpHeader] = Nil)(implicit decoder: Decoder[B]): FS[Response[B]] = toInj28(GetOP[B](uri, headers, decoder))
+      override def postUrlencoded[A, B](uri: Uri, payload: A, headers: List[HttpHeader] = Nil)(implicit encoder: UrlEncoded[A], decoder: Decoder[B]): FS[Response[B]] = toInj28(PostUrlencodedOP[A, B](uri, payload, headers, encoder, decoder))
     }
     implicit def to[LL$26[_]](implicit ii$27: _root_.freestyle.InjK[Op, LL$26]): To[LL$26] = new To[LL$26]
     def apply[LL$26[_]](implicit ev$29: JsonHttpClient[LL$26]): JsonHttpClient[LL$26] = ev$29
```

**Comment by fommil** [comment]
when flattening the `case class` parameter list, the `implicit` keyword should be removed. That deals with the warning... not so sure about the error.

**Comment by raulraja** [comment]
Seems like a change in behavior from scalamacro to scalameta where context bound are not treated as implicit args. So @fommil can't compile it right now but does it compile when you remove the context bound and add them as implicit params?

**Comment by fommil** [comment]
nope, that's the second case above.

**Comment by raulraja** [comment]
implicit args have worked before by whatever you pass implicit in the algebra the handlers receives as explicit args.

**Comment by fommil** [comment]
I think it might just be because `implicit` is being added as a keyword to the `case class` constructor... so `GetOP[B](uri, headers, decoder)` is failing because it should really be `GetOP[B]()(uri, headers, decoder)` when the compiler adds the non-implicit params

**Comment by fommil** [comment]
I think there may be two bugs in here... one to do with implicits not being propogated correctly and another to do with handling implicit parameter blocks.

**Comment by fommil** [comment]
hmm, yeah I'm confused how in `apply` the type `B` is visible such that `get` can be called appropriately. Seems that it should be encoded in `AA$31` is encoding `Response[B]`. But that is defined at construction time, I don't see how we could get a different `B` for each invocation of the method in that case.

**Comment by fommil** [comment]
Aha!

```
        case 0 =>
          val fresh33: GetOP[Any] = fa$32.asInstanceOf[GetOP[Any]]
          get(fresh33.uri, fresh33.headers)
```

should be

```
        case 0 =>
          val fresh33: GetOP[Any] = fa$32.asInstanceOf[GetOP[Any]]
          get(fresh33.uri, fresh33.headers, fresh33.<the implicit here>)
```

which I'm not sure how to do when using context bounds.

**Comment by fommil** [comment]
(and for the other case, the `implicit` keyword should be stripped from the `case class` and I *think* it should start to work)

**Comment by raulraja** [comment]
we can summon values from context bounds with `implicitly[Decoder[A]]`

**Comment by fommil** [comment]
it would be nicer if `scala.meta` just handled it the same as the explicit implicit case...

**Comment by fommil** [comment]
I don't think we can use `implicitly[Decoder[A]]`... the instance is sitting on the `case class` using an anonymously generated field name. Maybe `import fresh33._` before calling the method? (which would introduce weird shadowing bugs if any field names matched the method name...)

**Comment by raulraja** [comment]
We may be able to just summon the value and add it to the case class as a regular param so it's treated as such in the generated handler

**Comment by fommil** [comment]
cool, although I'm happy using explicit implicits as a workaround.

**Comment by fommil** [comment]
if you know how I can knock out that `implicit` keyword from the case class generation, I could have a go locally and see if it fixes it?

**Comment by raulraja** [comment]
Let me know if you need help, I'll be around

**Comment by raulraja** [comment]
we'd need to inspect the params and recreate them discarding the implicit mod

**Comment by fommil** [comment]
I basically fixed everything that was bugging me **except** this LOL https://github.com/frees-io/freestyle/pull/366

**Comment by xeno-by** [comment]
@fommil The lack of accurate positions is a limitation of scalameta/paradise that will be fixed in scalamacros/scalamacros.
